DOCSP-8058 NodeJS Connection Guide

ccho-mongodb opened the pull request:
JIRA: https://jira.mongodb.org/browse/DOCSP-8058

Staging:
https://docs-mongodbcom-staging.corp.mongodb.com/node/docsworker/DOCSP-8058-nodejs-connection-guide/getting-started/authentication

terakilobyte commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,412 @@
+   const { MongoClient } = require('mongodb');
+
+   // specify the placeholder values for your environment in the following line
+   const usernanme = encodeURIComponent(<username>);
```
```suggestion
   const username = encodeURIComponent(<username>);
```

terakilobyte commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,412 @@
+   async function run() {
+     try {
+       // Connect the client to the server
+       await client.connect();
```
So, the node driver has anomalous behavior that isn't spec compliant like the other drivers. I suggest adding in a `ping` command, which other drivers will need to do to trigger going from a lazy to active connection.

terakilobyte commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,412 @@
+     try {
+       // Connect the client to the server
+       await client.connect();
+       console.log('Connected successfully to server');
```
```suggestion
       await client.db('admin').command({ ping: 1 });
       console.log('Connected successfully to server');
```

terakilobyte commented:
@ccho-mongodb Did a first pass. It looks like you should globally find/replace `usernanme` with `username`, and I'd suggest adding in the ping command since in the future connect won't trigger auth.

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+appropriate to your user.
+
+For instructions on how to set up a new user or user permissions on your
+MongoDB server, see the **Authentication** and **Authorization** sections in
```
Here, I intentionally wanted to avoid linking out to other pages that were related to DBA-role responsibilities, but also wanted to avoid user confusion as to how permissions are set up. I also did not want to make any assumptions as to what environment they're working in -- whether they are developing locally or are attempting to connect to a managed instance. If it's confusing, I can omit it.

Also, we do not have the `extlinks` module (enables you to define shortcuts like the one you mentioned) in docs-node, so I'm not able to use that feature.

schmalliso replied:
The external links are built-in to snooty! https://github.com/mongodb/snooty-parser/blob/master/snooty/rstspec.toml#L602

ccho-mongodb replied:
Oh, darn. Thanks for that info. I had changed all the `:manual:` instances to external links. ☹️

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+When you connect to a MongoDB server, you provide your credentials and
+specific connection mechanism to the driver. The driver negotiates with the
+server to verify your identity, and the server grants you permissions
+appropriate to your user.
```
Thanks for pointing out it is confusing. I'll update this. I meant to describe what happens in the auth mechanism which is that you can now execute operations without receiving an error, and not to describe two separate topics (driver negotiation with server and setting permissions in the db).

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+Specify the ``DEFAULT`` option to the driver in order to authenticate with the
+server using a username and password. This option is a fallback setting that
+instructs the driver to negotiate the first authentication mechanism supported
+by the server using the following order of preference:
```
Thanks, will fix.

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+This guide shows you how to connect to MongoDB with the Node.js driver using
+different connection mechanisms as well as how to configure them.
+
+When you connect to a MongoDB server, you provide your credentials and
```
Thanks for that info! I can definitely use **instance** or **deployment**. I just wanted to be specific to avoid confusion between client/server and instance/deployment works fine.

schmalliso replied:
FAIR! People have historically had both client/server confusion and (mongodb) server / (host) server confusion, hence the instance/deployment preference, but it's just convention at this point :)

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+Overview
+--------
+
+This guide shows you how to connect to MongoDB with the Node.js driver using
```
1) Whoops, good catch. Will fix.

2) Will fix to make it more clear that they are common options. I actually mentioned to Nathan this morning that I was concerned the guide contained at least two different contexts: auth mechanisms and connection options, and that maybe it would be better to split them. I could even picture each auth mechanism on its own page since it's not common that a reader will need the context from more than one mechanism.

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+
+For instructions on how to set up a new user or user permissions on your
+MongoDB server, see the **Authentication** and **Authorization** sections in
+the `Security <https://docs.mongodb.com/manual/security/>`_ server manual.
```
Thanks, will fix.

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+~~~~~~~
+
+Specify the ``DEFAULT`` option to the driver in order to authenticate with the
+server using a username and password. This option is a fallback setting that
```
Thanks, I like that suggestion.

Side note: I'd like some input on what preposition to use with **authentication**. I've always used "with", but have seen "to" and "against". What's your opinion? I'm happy to go through and update all instances in this PR.

schmalliso replied:
Elsewhere (i.e. the manual) we mostly use "authenticate to". The pattern is you authenticate to MongoDB with your credentials.

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+server does not support that mechanism either, the driver attempts to
+authenticate using ``MONGODB-CR``.
+
+You can specify this option by omitting the ``authMechanism`` parameter of the
```
I deliberated this with myself for some time. Do you think we should show the `authMechanism=DEFAULT` since it does not need to be specified? Or, does it provide potential value by explicitly including it?

schmalliso replied:
I think there's value in showing it and also in mentioning you can leave it off because it's the default. Maybe show a connection string with it, and then say "since DEFAULT is the default auth mechanism, you can also leave off the authMechanism parameter entirely and MongoDB will blabla" -- but like a better sentence than that.

I think that the current wording of "you can specify DEFAULT by omitting the parameter" is currently confusing though.

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+see the `SCRAM <https://docs.mongodb.com/manual/core/security-scram/>`_ section
+in the server manual.
+
+SCRAM-SHA-256
```
Could you explain what this means? Did you mean use "``"? Since this is a sub-heading, I'm not sure that works.

schmalliso replied:
Sorry, yes, I do mean using ``foo`` - subheadings do (or at least historically have) supported being backticked, but I'm not married to it!

ccho-mongodb replied:
Oh, I see. I didn't know you could apply both formats. Will update.

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+   ``SCRAM-SHA-256`` is the default authentication method for MongoDB starting
+   in version 4.0
+
+Specify the ``SCRAM-SHA-256`` option to the driver in order to authenticate
```
Do you think it would be better to combine the first and second paragraph? Or, perhaps to leave out "specify X option" and instead describe what SCRAM-SHA-256 (or other connection mechanism) since what you're saying is restated in the second paragraph? E.g.

"The ``SCRAM-SHA-256`` authentication mechanism is a salted challenge-response mechanism (SCRAM) that uses your username and password, encrypted with the `SHA-256` algorithm to authenticate your user."

schmalliso replied:
I like this paragraph!

ccho-mongodb commented on source/getting-started/connection.txt:
```diff
@@ -0,0 +1,622 @@
+---------------------
+
+In this section, you can find descriptions and sample code for connecting to
+a MongoDB server using each of the connection mechanism settings.
```
Do you think it's better to leave out "descriptions"? I wanted to make it clear that they would be provided some idea of what they are as well.

schmalliso replied:
I'm cool with keeping "descriptions"!

ccho-mongodb commented:
I think I addressed everything in the latest commit @schmalliso , ready for re-review when time allows.

ccho-mongodb commented:
I updated the structure to separate community and enterprise auth mechanisms, removed the connection options (for now), and trimmed some of the info that seemed extraneous for the Lucas persona. Let me know what you think when you get the chance @terakilobyte @schmalliso
